File: src/main/java/ru/spb/se/contexthelper/ui/StackExchangeThreadsTree.java
package ru.spb.se.contexthelper.ui;

import com.google.code.stackexchange.schema.Answer;
import com.google.code.stackexchange.schema.Question;
import javax.swing.JTree;

/** {@link JTree} view component for displaying StackExchange threads. */
class StackExchangeThreadsTree extends JTree {

  StackExchangeThreadsTree(
      StackExchangeTreeListener treeListener, StackExchangeThreadsTreeModel treeModel) {
    super(treeModel);
    setCellRenderer(new StackExchangeThreadsTreeCellRender());
    setRootVisible(false);
    addTreeSelectionListener(e -> {
      Object lastSelectedComponent = getLastSelectedPathComponent();
      if (lastSelectedComponent instanceof Question) {
        Question question = (Question) lastSelectedComponent;
        //TODO: Add question.link as a first line of the text.
        treeListener.questionClicked(question);
        treeListener.renderPrettifiedHtml(question.getBody());
      } else if (lastSelectedComponent instanceof Answer) {
        Answer answer = (Answer) lastSelectedComponent;
        treeListener.answerClicked(answer);
        treeListener.renderPrettifiedHtml(answer.getBody());
      }
    });
  }
}

File: src/main/java/ru/spb/se/contexthelper/ui/StackExchangeThreadsTree.kt
package ru.spb.se.contexthelper.ui

import com.google.code.stackexchange.schema.Answer
import com.google.code.stackexchange.schema.Comment
import com.google.code.stackexchange.schema.Question
import javax.swing.JTree
import javax.swing.tree.TreeModel

/** {@link JTree} view component for displaying StackExchange threads. */
class StackExchangeThreadsTree(
    treeListener: StackExchangeTreeListener,
    treeModel: TreeModel
): JTree(treeModel) {
    init {
        setCellRenderer(StackExchangeThreadsTreeCellRender())
        isRootVisible = false
        addTreeSelectionListener { _ ->
            val component = lastSelectedPathComponent
            when (component) {
                is Question -> {
                    treeListener.questionClicked(component)
                    treeListener.renderPrettifiedHtml(component.body)
                }
                is Answer -> {
                    treeListener.answerClicked(component)
                    treeListener.renderPrettifiedHtml(component.body)
                }
                is Comment -> {
                    treeListener.renderPrettifiedHtml(component.body)
                }
            }
        }
    }
}